Refer to groups by UUID

orange-elephant opened the pull request:
Happy new year!

For now entries can only be added to 1 group, but basically everything needed to make adding to multiple possible is already laid down here, and I'll be happy to work on that if this PR is merged.

alexbakker commented:
Awesome work as always! Unfortunately I haven't had time to look at this yet, and I may not be able to get to it in the next couple of weeks either. Sorry about that. Starting February I'll should have some more time for Aegis.

alexbakker commented on app/src/main/java/com/beemdevelopment/aegis/vault/Vault.java:
```diff
@@ -49,17 +70,51 @@ public static Vault fromJson(JSONObject obj) throws VaultException {
                 entries.add(entry);
             }
+
+            if (obj.has("groups")) {
```
Similarly to the importer, I think we should parse the groups first, then parse the entries and then check the groups they're in. With regards to the entries: I think we may have to move the check for the old ``group`` field here. The reason being: This is the only place where we can actually do the conversion to the new group system. Now we first add the entry and only then we check whether we need to convert it.

_(Having to do all this referential integrity checking manually really makes you wish we were using SQLite)_

alexbakker reviewed (changes requested):
Apologies for the delay. I've added some initial comments. Don't be afraid to push back if you think I'm asking for something that wouldn't work well. With a large patch like this it's easy to miss details.

orange-elephant commented on app/src/main/java/com/beemdevelopment/aegis/vault/VaultGroup.java:
```diff
@@ -0,0 +1,91 @@
+        return new VaultGroup(NO_GROUP, context.getString(R.string.no_group));
+    }
+
+    public static UUID getNoGroupUuid() {
```
The UI placeholder for 'no group' can be removed entirely once multiple groups support is added since it will be a dropdown of checkboxes and checking 0 would be equivalent. This was the plan but I felt doing it all in one PR would be too big of a change.

I've seen this approach in other codebases to put placeholder items into ViewHolders easily and having a VaultGroup object for 'no group' makes using your dropdown helpers easy. Its up to you whether you think its is too hacky to use for now. It shouldn't be a problem to just use null for filtering (which I agree seems more logical).

I had considered that a magic UUID like this could be used for things like a recycle bin or archive feature in future.

orange-elephant commented:
I've finally gotten round to addressing the feedback, everything you suggested was okay. The only thing I'm still not sure about is how to get the "New group" button into the dropdown checkboxes so I've reverted that to the original approach for now and see if I can come up with a workaround.

It would probably be possible to migrate entries that claim to be in a group that has been discarded due to duplicate names to the UUID of the group of the same name which was kept but this would be a super rare circumstance and the user could easily add them back to the group so I'm not sure if that would be worth it, what do you think?

orange-elephant commented:
> why we need to call `Vault.migrateOldGroupIfNecessary(_vault, entry);` in `VaultRepository.addEntry(VaultEntry entry)`.

This was because `addEntry` is used instead of `Vault.fromJson` when importing a JSON vault file (at least anytime other than the intro slides) so if the JSON backup had an old style group then this needs to be converted when it is added to the vault. This conversion may only need to happen if the entry is being added from ImportEntriesActivity since I don't think its possible for QR codes to contain info about a group.

> Also, I've made a couple of minor changes here and there, I hope you don't mind.

No problem, the group model is much better than what I had

alexbakker reviewed (approved):
Thanks a lot!
